src/cryptonote_basic/atomic_lock.h

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2014-2018, The Monero Project
+// 
+// All rights reserved.
+// 
+// Redistribution and use in source and binary forms, with or without modification, are
+// permitted provided that the following conditions are met:
+// 
+// 1. Redistributions of source code must retain the above copyright notice, this list of
+//    conditions and the following disclaimer.
+// 
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+// 
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <type_traits>
+#include <atomic>
+
+
+
+namespace cryptonote
+{
+  template <class T>
+  class atomic_lock
+  {
+    public:
+      // the type of the underlying storage for
+      // the enum that we are switching between
+      using storage_type = typename std::underlying_type<T>::type;
+
+      atomic_lock() = default;
+      atomic_lock(T value) : m_value{ static_cast<storage_type>(value) } {}
+
+      /**
+       * Check whether we currently hold a specific lock
+       * @param     value   The value to compare to
+       * @return    Do we currently hold the given value
+       */
+      bool is(T value) const noexcept {
+        return m_value.load() == static_cast<storage_type>(value);
+      }
+
+      /**
+       * Try to advance to a new value, fails if we were not at
+       * the given starting point to begin with
+       * @param     from    The point to advance from
+       * @param     to      The point to advance to
+       * @return    Did we successfully advance
+       */
+      bool try_advance(T from, T to) noexcept
+      {
+        // check if the original value is correct
+        if (m_value.fetch_add(difference(from, to)) != static_cast<storage_type>(from)) {
+          // we did not start at from so undo the add
+          m_value.fetch_sub(difference(from, to));
+          return false;
+        }
```

Comment:
That's not atomic.

Reply:
You are right, but it's not relevant here. The point is that when two threads try to change the value, only one of them will get the okay. The other thread will see the already-changed value and then revert its change. This ensures thread safety.

Reply:
How about:

thread 1 calls this, value goes up
thread 2 calls this, value goes up, but doesn't run the second call yet
thread 1 calls the other, values goes down
thread2 calls the second call

Reply:
Right, that is a possible scenario, and it would work just fine. Let's assume the value used for (in/de)crementing is 1 and we start with a value of 0. We'd get the following result activity

Thread 1 ups the value: 0 => 1. Old value was 0 so it continues
Thread 2 ups the value: 1 => 2. Old value was 1 so it does _not_ continue
Thread 1 downs the value: 2 => 1
Thread 2 aborts, downs the value: 1 => 0

Where in this scenario would there be a problematic race-condition?

Reply:
The value is temporarily set to a different value, even if the call fails. The is function will not return what's expected.

Reply:
You are right. This should have been done with compare_exchange instead, I wasn't even aware of this functionality. it makes the code easier as well.